include/fi_log.h

```diff
@@ -35,32 +35,137 @@
 #define FI_LOG_H
 
 #if HAVE_CONFIG_H
-#  include <config.h>
+#include <config.h>
 #endif /* HAVE_CONFIG_H */
 
-extern int fi_log_level;
+#include <fi_list.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+/*
+ * Define the basic subsystems that can be used to enable fine grained logging.
+ */
+#define FABRIC (0)
+#define DOMAIN (1)
+#define EP_CM (2)
+#define EP_DM (3)
+#define AV (4)
+#define CQ (5)
+#define EQ (6)
+#define MR (7)
+
+#define LOG_WARN (1)
+#define LOG_TRACE (3)
+#define LOG_INFO (7)
+
+/*
+ * Log level occupies bit positions 0-2
+ * Subsystem occupies bit positions 3-31
+ * Provider occupes bit positions 32-63
+ */
+#define SUBSYS_OFFSET 3
+#define PROV_OFFSET 32
+
+#define MASK(width) ((UINT64_C(1) << (width)) - 1)
+
+#define SUBSYS_MASK (MASK(29) << SUBSYS_OFFSET)
+#define PROV_MASK (MASK(32) << PROV_OFFSET)
+
+/*
+ * Take bit position and offset and set bit in UINT64
+ */
+#define EXPAND(position, offset) (UINT64_C(1) << ((position) + (offset)))
+
+extern uint64_t log_mask;
+
+struct log_providers {
+	struct slist names;
+	bool negated;
+};
+
+struct provider_parameter {
+	struct slist_entry entry;
+	char *name;
+};
+
+struct prov_handle {
+	struct slist_entry entry;
+	char *name;
+	char bitpos;
+};
 
 void fi_log_init(void);
-void fi_warn_impl(const char *prov, const char *fmt, ...);
-void fi_log_impl(int level, const char *prov, const char *func, int line,
-		 const char *fmt, ...);
-void fi_debug_impl(const char *prov, const char *func, int line, const char *fmt, ...);
+struct prov_handle *fi_log_register_provider(const char *prov_name);
+
+
+void fi_err_impl(struct prov_handle *prov, const char *fmt, ...);
+void fi_serr_impl(struct prov_handle *prov, int subsystem, const char *fmt,
+		...);
+void fi_log_impl(struct prov_handle *prov, int level, const char *func, int
+		line, const char *fmt, ...);
+void fi_slog_impl(struct prov_handle *prov, int level, int subsystem, const
+		char *func, int line, const char *fmt, ...);
+void fi_debug_impl(struct prov_handle *prov, const char *func, int line, const
+		char *fmt, ...);
+void fi_sdebug_impl(struct prov_handle *prov, int subsystem, const char *func,
+		int line, const char *fmt, ...);
+
```

Comment:
I don't really see the need to maintain duplicate calls just to add the subsystem parameter.